k2/csrc/array.h

```diff
@@ -1,69 +1,133 @@
 // k2/csrc/array.h
 
 // Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, I> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
+  using T = typename std::remove_pointer<Ptr>::type;
+
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes.
+
+  operator T *() { return data; }
+  operator const T *() const { return data; }
+
+  // ++it
+  StridedPtr &operator++() {
+    data += stride;
+    return *this;
+  }
+
+  // it++
+  StridedPtr operator++(int) {
+    StridedPtr tmp(*this);
+    data += stride;
+    return tmp;
+  }
+
+  bool operator==(const StridedPtr &other) const { return data == other.data; }
+
   StridedPtr(T *data, I stride) : data(data), stride(stride) {}
 };
 
-/* MIGHT NOT NEED THIS */
-template <typename I, typename Ptr>
-struct Array1 {
-  // Irregular two dimensional array of something, like vector<vector<X> >
+template <typename Ptr, typename I = int32_t>
+class Array1 {
+ public:
+  // Irregular two dimensional array of something, like vector<X>
   // where Ptr is, or behaves like, X*.
   using IndexT = I;
   using PtrT = Ptr;
+  using ValueType = typename std::remove_pointer<PtrT>::type;
+  using PtrWithStride = StridedPtr<Ptr, I>;
```

Comment:
I think we should not use any `stride` here,  we could just pass StridedPtr as `Ptr` in the template parameter, then let `operator[]`(you deleted above) in `StridedPtr` do everything else. `end` or `begin` just return the positions (`end - begin` is the total number of elements), `data[i]` will call StridedPtr.data[i], which is `stridedPtr.data[i*stride]`

Reply:
`stride` is not a compile-time constant, how do you let `StridePtr` know what the stride it has?

Reply:
The working pipline may be as below if we want provide StridedPtr as Ptr in Array:

1. Construct StridedPtr with data and stride.
2. Create Array with Array1<StridedPtr, I> or Array2<StridedPtr, I>.
3. Access `data[i]` in Array, it calls `Ptr.data[i]` which is `StridedPtr.data[i]`

Reply:
The main idea here is that `array` is just an interface to access data[i], it does not care about the storage structure (StridedPtr cares about this).

Reply:
@qindazhu is right here, I think.
The idea with StridedPtr is you know at compile time that there is a stride that is not 1, but it gets passed in via the StridedPtr.

Reply:
If the stride is known at compile time, does it mean we should instantiate the template
for a given list of strides and choose which one to use at the runtime?